File: original_version/0.expected_error.py
import sys
import gzip
if len(sys.argv) < 3 :
	print("Usage: python " + sys.argv[0] + " [raw.fastq] " + "[expected error threshod]")
	exit()

filename = sys.argv[1]
threshod = int(sys.argv[2])
outfile = filename + ".clean.fq"


#Calculate phred score.
def exp_e(q):
	lenq = len(q)
	exp = 0
	phred = 0
	tmp = list(q)
	ascill = [ ord(n) - 33 for n in tmp ]

	for i in ascill:
		exp += 10 **(-i/10)

	return exp
#---------------------------------------------------------
#Read sequences.
err = open("err.fastq",'w')
log = open("log.txt",'w')
out = open(outfile,'w')

total = 0
nn = 0
clean = 0

with gzip.open(filename,'rt') as fh:
	id = fh.readline().strip()
	if id[0] != '@':
		print("ERROR: {} is not a correct fastq format".format(filename))
		exit()
	while id:
		seq = fh.readline().strip()
		fh.readline().strip()
		qual = fh.readline().strip()
		if 'N' in seq:
			err.write(id + "\n" + seq + "\n" + "+\n" + qual + "\n")
			nn += 1
	
		else:
			if exp_e(qual) < threshod :
				out.write(id + "\n" + seq + "\n" + "+\n" + qual + "\n" )
				clean += 1
		id = fh.readline().strip()

err.write("total reads:\t{}".format(total))
err.write("contain Ns reads:\t{}".format(nn))
err.write("clean reads:\t{}".format(clean))
err.close()
log.close()
out.close()

File: original_version/0.expected_error.pl

#!/usr/bin/perl -w
use strict;
unless (@ARGV >0) {
	print "usage:perl $0 <*.fastq>";
	exit;
}
my $filename = shift;

	if ($filename =~ /.gz$/) 
{
	open FQ,"gzip -dc $filename|" or die "Can not open $filename:$!\n";
}else{
	open FQ,"$filename" or die "Can not open $filename:$!\n";
}

#Read sequences.

while(my $t1=<FQ>) {
	chomp($t1);
	if ($t1 !~ m/^@/)
	{
		print STDERR "ERROR:\n$filename: It's not a correct fastq format.\n";
		exit;
	}
	$t1 =~s/^@//;
	chomp(my $seq=<FQ>);
	<FQ>;
	chomp(my $qual=<FQ>);
	my $l = length $seq;
	my $ql = length $qual;

	my ($aver_q,$aver_e) = &cal_ascii($qual);
	print "$t1\t$aver_q\t$aver_e\n";
}
close FQ;

#Calculate phred score.
sub cal_ascii{
	my $ascii_str = shift ;
	my $len = length $ascii_str;
	my ($exp_e,$phred) = (0,0);
	chomp $ascii_str;
	my @tmp = split //,$ascii_str;
	my @all_ascii = map {ord($_) - 33} @tmp;
	for my $i(0..$#all_ascii){
		$exp_e += 10 **(-$all_ascii[$i]/10);
		$phred += $all_ascii[$i];

	}
	#my $average_e = sprintf("%.2f",$exp_e/$len);
	my $average_q = sprintf("%.2f",$phred/$len);

	return ($average_q,$exp_e);
}

